File: mcp-brokered-credentials-typescript/src/tools/execute.ts
import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import type { AuthProvider } from "@keycardai/mcp/server/auth/provider";
import { z } from "zod";
import { withLinearClient } from "../upstream.js";

export function registerExecuteTool(
  server: McpServer,
  authProvider: AuthProvider,
) {
  server.tool(
    "execute",
    "Execute a tool on the upstream Linear MCP server. Use `search` first to discover available tool names and their input schemas.",
    {
      name: z
        .string()
        .min(1)
        .describe("Upstream tool name, exactly as returned by `search`."),
      arguments: z
        .record(z.unknown())
        .optional()
        .describe(
          "Arguments object matching the upstream tool's inputSchema. Omit for no-argument tools.",
        ),
    },
    async ({ name, arguments: args }, extra) => {
      const subjectToken = extra.authInfo?.token;
      if (!subjectToken) {
        throw new Error("Missing bearer token on the incoming request.");
      }

      const result = await withLinearClient(
        authProvider,
        subjectToken,
        (client) =>
          client.callTool({
            name,
            arguments: args ?? {},
          }),
      );

      // The upstream may return the legacy `{ toolResult }` shape; normalise
      // to the modern `{ content }` shape so callers always see the same
      // response structure.
      const content = Array.isArray(result.content)
        ? result.content
        : [
            {
              type: "text" as const,
              text: JSON.stringify(result.toolResult ?? result, null, 2),
            },
          ];

      return {
        content,
        ...(result.structuredContent !== undefined && {
          structuredContent: result.structuredContent as Record<string, unknown>,
        }),
        ...(result.isError !== undefined && { isError: Boolean(result.isError) }),
      };
    },
  );
}

File: mcp-brokered-credentials-python/tools/execute.py
import json
from typing import Any

from mcp.server.fastmcp import Context, FastMCP
from keycardai.mcp.server.auth import AuthProvider

from upstream import linear_client_for_user


def register_execute_tool(mcp: FastMCP, auth_provider: AuthProvider) -> None:
    @mcp.tool(
        description=(
            "Execute a tool on the upstream Linear MCP server. "
            "Use `search` first to discover available tool names and their input schemas."
        )
    )
    async def execute(
        ctx: Context,
        name: str,
        arguments: dict[str, Any] | None = None,
    ) -> str:
        """name: Tool name exactly as returned by search. arguments: Input matching the tool's inputSchema."""
        async with linear_client_for_user(auth_provider, ctx) as client:
            result = await client.call_tool(name, arguments=arguments or {})

            content = [
                {"type": c.type, "text": c.text if hasattr(c, "text") else str(c)}
                for c in (result.content or [])
            ]
            payload: dict[str, Any] = {"content": content}
            if result.isError:
                payload["isError"] = True
            return json.dumps(payload, indent=2)
